src/plot.jl

```diff
@@ -1,164 +1,110 @@
-export MeanPlot, AutocorPlot, HistogramPlot, DensityPlot, MixedDensityPlot, TracePlot
-export plot
-
-@userplot MeanPlot
-@userplot AutocorPlot
-@userplot HistogramPlot
-@userplot DensityPlot
-@userplot MixedDensityPlot
-@userplot TracePlot
-
-supportedplots = [
-                  MeanPlot, 
-                  AutocorPlot, 
-                  HistogramPlot,
-                  DensityPlot,
-                  MixedDensityPlot,
-                  TracePlot
-]
-colorparamters = [:chain, :parameter]
-
-@recipe function f(p::MeanPlot; colordim = :chain)
-    c, i = p.args
-    seriestype := :line
-    xaxis := "Iteration"
-    yaxis := "Mean"
-    
-    @assert colordim ∈ colorparamters
+
+@shorthands meanplot
+@shorthands autocorplot
+@shorthands mixeddensity
+@shorthands traceplot
+
+struct _TracePlot; c; val; end
+struct _MeanPlot; c; val;  end
+struct _DensityPlot; c; val;  end
+struct _HistogramPlot; c; val;  end
+struct _AutocorPlot; lags; val;  end
+
+# define alias functions for old syntax
+const translationdict = Dict(
+                        :traceplot => _TracePlot,
+                        :meanplot => _MeanPlot,
+                        :density => _DensityPlot,
+                        :histogram => _HistogramPlot,
+                        :autocorplot => _AutocorPlot
+                      )
+
+const supportedplots = push!(collect(keys(translationdict)), :mixeddensity)
+
+@recipe function f(c::AbstractChains, i::Int; colordim = :chain, barbounds = (0, Inf), maxlag = nothing)
+    st = get(plotattributes, :seriestype, :traceplot)
 
     if colordim == :parameter
         title := "Chain $(c.chains[i])"
         labels := c.names
-        c.range, MCMCChain.cummean(c.value[:, :, i])
-    else
+        val = c.value[:, :, i]
+    elseif colordim == :chain
         title := c.names[i]
         labels := map(k -> "Chain $(c.chains[k])", 1:size(c)[3])
-        c.range, MCMCChain.cummean(c.value[:, i, :])
+        val = c.value[:, i, :]
+    else
+        throw(ArgumentError("`colordim` must be one of `:chain` or `:parameter`"))
     end
-end
 
-@recipe function f(p::AutocorPlot; maxlag = nothing, colordim = :chain)
-    c, i = p.args
-    lags = 0:(maxlag === nothing ? round(Int, 10 * log10(length(c.range))) : maxlag)
-    ac = MCMCChain.autocor(c, lags=collect(lags))
-    xaxis := "Lag"
-    yaxis := "Autocorrelation"
+    if st == :mixeddensity
+        discrete = MCMCChain.indiscretesupport(c, barbounds)
+        st = (discrete[i] && colordim == :chain) ? :histogram : :density
+        seriestype := st
+    end
 
-    @assert colordim ∈ colorparamters
-    if colordim == :parameter
-        title := "Chain $(c.chains[i])"
-        labels := c.names
-        lags, ac.value[:, :, i]'
+    if st == :autocorplot
```

Comment:
unfortunately special casing is needed for autocorplot, it's not so elegant but hey